### PR TITLE
[core] Fix all clang-tidy performance-* errors

### PR DESCRIPTION
**Pull request by tmpsantos**
Also make the bot fail if they come back. In theory we should reduce the number of heap allocations.

Warnings fixed by this PR:

```
      2 [performance-inefficient-vector-operation]
      2 [performance-noexcept-move-constructor]
      8 [performance-for-range-copy]
     19 [performance-type-promotion-in-math-fn]
    212 [performance-unnecessary-value-param]
```

- [x] Measure improvements in heap allocation before merging this PR.

### CODE REVIEW

**Comment by tmpsantos** [comment]
Interesting results, when running the render tests, it consistently saves 200k memory allocations for a full run:
image

When I was trying to find out where these savings were coming from, I notice that some results are misleading in the tool, for instance it some cases it will move allocations down/up in the stack frame.

These are the two functions contributing to most of the savings:
image

Interestingly, this was the only change in `BiDi::processText`:

```
@@ -98,7 +99,7 @@ std::vector<std::u16string> BiDi::processText(const std::u16string& input,
         throw std::runtime_error(std::string("BiDi::processText: ") + u_errorName(errorCode));
     }
 
-    return applyLineBreaking(lineBreakPoints);
+    return applyLineBreaking(std::move(lineBreakPoints));
 }
```

The other functions showing are not really improvements they show up again at the bottom as regression, canceling each other. Reason is the function signature changed or allocation moved to some other frame.

**Comment by tobrun** [comment]
tenor

File: src/mbgl/actor/mailbox.cpp [hunk 1]
```diff
@@ -91,14 +91,14 @@ void Mailbox::receive() {
...
+void Mailbox::maybeReceive(const std::weak_ptr<Mailbox>& mailbox) {
     if (auto locked = mailbox.lock()) {
         locked->receive();
     }
 }
 
 // static
-std::function<void()> Mailbox::makeClosure(std::weak_ptr<Mailbox> mailbox) {
+std::function<void()> Mailbox::makeClosure(const std::weak_ptr<Mailbox>& mailbox) {
     return [mailbox]() { maybeReceive(mailbox); };
```
Comment (alexshalamov) [inline src/mbgl/actor/mailbox.cpp]: maybe we could have moved mailbox instead of copying it?

Comment (tmpsantos) [reply src/mbgl/actor/mailbox.cpp]: Done.

**Pull request merged**